Rectangular concrete linear map of Hyperplane/Line

`linear_map(M, P)` for `P` being a `Hyperplane` or a `Line` always creates a set of the same type. But this is not correct if `M` is not square (i.e., has different output dimension).
```
julia> M = [1.0 1.0]
1×2 Array{Float64,2}:
 1.0  1.0

julia> linear_map(M, rand(Line))  # tries to create a Line but then fails
ERROR: AssertionError: lines must be two-dimensional

julia> H = rand(Hyperplane)
Hyperplane{Float64}([-0.1514314278974606, -0.3351825271100233], 1.2796222453508572)

julia> linear_map(M, H)  # should be the one-dimensional universe
Hyperplane{Float64}([-0.24330697750374183], 1.2796222453508572)
```

--- a/src/Sets/HalfSpace.jl
+++ b/src/Sets/HalfSpace.jl
@@ -487,7 +487,7 @@ function _linear_map_hrep_helper(M::AbstractMatrix{N}, P::HalfSpace{N},
     if length(constraints) == 1
         return first(constraints)
     elseif isempty(constraints)
-        return Universe(size(M, 1))
+        return Universe{N}(size(M, 1))
     else
         return HPolyhedron(constraints)
     end

--- a/src/Sets/HalfSpace.jl
+++ b/src/Sets/HalfSpace.jl
@@ -486,6 +486,8 @@ function _linear_map_hrep_helper(M::AbstractMatrix{N}, P::HalfSpace{N},
     constraints = _linear_map_hrep(M, P, algo)
     if length(constraints) == 1
         return first(constraints)
+    elseif isempty(constraints)
+        return Universe(size(M, 1))
     else
         return HPolyhedron(constraints)
     end

--- a/src/Sets/Hyperplane.jl
+++ b/src/Sets/Hyperplane.jl
@@ -432,9 +432,15 @@ end
 function _linear_map_hrep_helper(M::AbstractMatrix{N}, P::Hyperplane{N},
                                  algo::AbstractLinearMapAlgorithm) where {N<:Real}
     constraints = _linear_map_hrep(M, P, algo)
-    @assert length(constraints) == 2
-    c = first(constraints)
-    return Hyperplane(c.a, c.b)
+    if length(constraints) == 2
+        # assuming these constaraints define a hyperplane
+        c = first(constraints)
+        return Hyperplane(c.a, c.b)
+    elseif isempty(constraints)
+        return Universe(size(M, 1))
+    else
+        error("unexpected number of $(length(constraints)) constraints")
+    end
 end
 
 """

--- a/src/Sets/Hyperplane.jl
+++ b/src/Sets/Hyperplane.jl
@@ -437,7 +437,7 @@ function _linear_map_hrep_helper(M::AbstractMatrix{N}, P::Hyperplane{N},
         c = first(constraints)
         return Hyperplane(c.a, c.b)
     elseif isempty(constraints)
-        return Universe(size(M, 1))
+        return Universe{N}(size(M, 1))
     else
         error("unexpected number of $(length(constraints)) constraints")
     end

--- a/src/Sets/Hyperplane.jl
+++ b/src/Sets/Hyperplane.jl
@@ -433,7 +433,7 @@ function _linear_map_hrep_helper(M::AbstractMatrix{N}, P::Hyperplane{N},
                                  algo::AbstractLinearMapAlgorithm) where {N<:Real}
     constraints = _linear_map_hrep(M, P, algo)
     if length(constraints) == 2
-        # assuming these constaraints define a hyperplane
+        # assuming these constraints define a hyperplane
         c = first(constraints)
         return Hyperplane(c.a, c.b)
     elseif isempty(constraints)

--- a/src/Sets/Line.jl
+++ b/src/Sets/Line.jl
@@ -322,10 +322,16 @@ end
 
 function _linear_map_hrep_helper(M::AbstractMatrix{N}, P::Line{N},
                                  algo::AbstractLinearMapAlgorithm) where {N<:Real}
-     constraints = _linear_map_hrep(M, P, algo)
-     @assert length(constraints) == 2
-     c = first(constraints)
-     return Line(c.a, c.b)
+    constraints = _linear_map_hrep(M, P, algo)
+    if length(constraints) == 2
+        # assuming these constraints define a line
+        c = first(constraints)
+        return Line(c.a, c.b)
+    elseif isempty(constraints)
+        return Universe(size(M, 1))
+    else
+        error("unexpected number of $(length(constraints)) constraints")
+    end
 end
 
 """

--- a/src/Sets/Line.jl
+++ b/src/Sets/Line.jl
@@ -328,7 +328,7 @@ function _linear_map_hrep_helper(M::AbstractMatrix{N}, P::Line{N},
         c = first(constraints)
         return Line(c.a, c.b)
     elseif isempty(constraints)
-        return Universe(size(M, 1))
+        return Universe{N}(size(M, 1))
     else
         error("unexpected number of $(length(constraints)) constraints")
     end

--- a/test/unit_HalfSpace.jl
+++ b/test/unit_HalfSpace.jl
@@ -111,6 +111,12 @@ for N in [Float64, Rational{Int}, Float32]
     M = N[2 2; 0 1] # invertible matrix
     @test linear_map(M, H) == HalfSpace(N[0.5, -2.0], N(0.0))
 
+    if test_suite_polyhedra
+        if N == Float64 || N == Float32
+            @test linear_map(N[1 1], H) == Universe{N}(1)
+        end
+    end
+
     # conversion of the normal vector
     hs_sev = HalfSpace(SingleEntryVector(2, 3, N(1)), N(1))
     hs_vec = convert(HalfSpace{N, Vector{N}}, hs_sev)

--- a/test/unit_Hyperplane.jl
+++ b/test/unit_Hyperplane.jl
@@ -73,6 +73,9 @@ for N in [Float64, Rational{Int}, Float32]
         if N == Float32 || N == Float64
             @test lm isa Hyperplane{Float64}
             @test lm.a ≈ N[0, -1] && lm.b ≈ N(0)
+
+            # returned set is universal
+            @test linear_map(N[1 1], H) == Universe{N}(1)
         elseif N == Rational{Int}
             @test lm isa Hyperplane{Rational{BigInt}}
             @test lm.a == N[0//1, -1//1] && lm.b == N(0//1)

--- a/test/unit_Line.jl
+++ b/test/unit_Line.jl
@@ -78,6 +78,9 @@ for N in [Float64, Rational{Int}, Float32]
         if N == Float32 || N == Float64
             @test lm isa Line{Float64}
             @test lm.a ≈ N[0, -1] && lm.b ≈ N(0)
+
+            # returned set is universal
+            @test linear_map(N[1 1], L) == Universe{N}(1)
         elseif N == Rational{Int}
             @test lm isa Line{Rational{BigInt}}
             @test lm.a == N[0//1, -1//1] && lm.b == N(0//1)

